### PR TITLE
safety check python packages

### PR DESCRIPTION
**Pull request by drageset**
Using the python package "safety" to check the python distribution for vulnerabilities. Triggers a build fail for pillow==8.3.1 as per DAPLA-486 https://statistics-norway.atlassian.net/jira/software/projects/DAPLA/boards/10?selectedIssue=DAPLA-486

I removed pillow from the build again, so we can merge safely.

### CODE REVIEW

**Comment by drageset** [comment]
It seems like Google Vulnerability scan only checks the OS, not the python packages. We need to find another way of scanning the python packages.

File: docker/jupyterlab/azure-pipelines.yml [hunk 1]
```diff
@@ -213,6 +221,14 @@ jobs:
             ${{ variables.imageTagLatest }}
             ${{ variables.imageTagBeforeVulnerabilityScan }}
 
+      # Trivy vulnerability scan, includes python packages installed in image
+      - task: CmdLine@2
+        displayName: "Trivy vulnerability scan final image"
+        inputs:
+          script: |
+            trivy image --exit-code 0 --severity LOW,MEDIUM ${{ variables.imageName }}:${{ variables.imageTagLatest }}
+            trivy image --exit-code 1 --severity HIGH,CRITICAL ${{ variables.imageName }}:${{ variables.imageTagLatest }}
```
Comment (arneso-ssb) [inline docker/jupyterlab/azure-pipelines.yml]: Vil foreslå at du legger til opsjonene `--light --ignore-unfixed`
--light gir mer kompakt output, og --ignore-unfixed gjør at bygget ikke feiler på sikkerhetshull som det ikke er publisert fiks for ennå. Det er kun de tilfellene hvor de finnes en fiks vi kan gjøre noe med det.

**Comment by drageset** [comment]
In the final version of the pipeline, the Trivy vulnerability check failed because there was no space left on device (EDIT: when Trivy tried to unpack an image to scan it, it could not). I think that might be due to the fact that several instances of the pipeline were running at once, and I can't cancel the other runs. We should try running the latest version of the pipeline again once all the runs have timed out of completed.

**Review by fenglin-ssb — APPROVED** [review]
LGTM

**Comment by drageset** [comment]
`2021-12-20T13:35:52.505Z	INFO	Need to update DB
2021-12-20T13:35:52.505Z	INFO	Downloading DB...
2.51 MiB / 6.88 MiB [---------------------->________________________________________] 36.48% ? p/s ?6.88 MiB / 6.88 MiB [-------------------------------------------------------------->] 99.98% ? p/s ?6.88 MiB / 6.88 MiB [-----------------------------------------------------] 100.00% 33.31 MiB p/s 1s2021-12-20T13:38:41.960Z	FATAL	scan error: image scan failed: failed analysis: analyze error: failed to analyze layer: sha256:3b12dc5ae360aa1ed5caae5f9afef3fa682f2f3be3c5acdd91de61f653ad72b0 : unable to get uncompressed layer sha256:3b12dc5ae360aa1ed5caae5f9afef3fa682f2f3be3c5acdd91de61f653ad72b0: failed to get the layer (sha256:3b12dc5ae360aa1ed5caae5f9afef3fa682f2f3be3c5acdd91de61f653ad72b0): unable to populate: unable to open: failed to copy the image: write /tmp/fanal-3773533808: no space left on device
2021-12-20T13:41:21.413Z	FATAL	scan error: image scan failed: failed analysis: analyze error: failed to analyze layer: sha256:3b12dc5ae360aa1ed5caae5f9afef3fa682f2f3be3c5acdd91de61f653ad72b0 : unable to get uncompressed layer sha256:3b12dc5ae360aa1ed5caae5f9afef3fa682f2f3be3c5acdd91de61f653ad72b0: failed to get the layer (sha256:3b12dc5ae360aa1ed5caae5f9afef3fa682f2f3be3c5acdd91de61f653ad72b0): unable to populate: unable to open: unable to export the image: Error response from daemon: write /var/lib/docker/tmp/docker-export-286548197/930c550180f2d893a15c262c0a05486d6ae8bf23c811616a16d36afdf2387d83/layer.tar: no space left on device
##[error]Bash exited with code '1'.`

This is output including the "no space left on device" error from the Trivy image scan step of the Azure Pipeline.

**Comment by drageset** [comment]
I have now moved the Trivy vulnerability scan to happen inside the image, with Docker RUN command. The Conda Python3.9 folder is the one that is being scanned, specifically.

**Comment by drageset** [comment]
trivy does not scan the python distribution by specifying its path with the filesystem command. We can use Trivy to uncover other image vulnerabilities, and use Python safety check to uncover vulnerable python packages.

**Comment by drageset** [comment]
We agreed that trivy should not automatically fail the build because it might detect vulns in jupyterlab base that are not "real" and we might be willing to live with, but safety check should fail, because we don't want to introduce more vulns through python package installs.

**Comment by drageset** [comment]
Now that safety check is included in the image, and it fails the build when it finds pillow, and pillow has been removed again, we are ready to merge the branch into master, as soon as the build completes without errors!

**Pull request merged**